tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+
+## Example: `Functor` deriving
+
+Of course, after talking so much about `Functor` instances and adding the
+clumsy `p` parameter to support them, we absolutely must to derive `Functor`
+instance for generics! Anyway, `Functor` instance for generics is already
+defined in `GHC.Generics`, so instead of re-implementing it, let's just go
+through the code:
+
+```haskell
+-- If we have a parameter ‘p’, we just map over it, as expected:
+
+instance Functor Par1 where
+  fmap f (Par1 p) = Par1 (f p)
+
+-- The same with ‘Rec1’ (just use ‘fmap’ because the inner part is a ‘Functor’):
+
+instance Functor f => Functor (Rec1 f) where
+  fmap f (Rec1 a) = Rec 1 (fmap f a)
+
+-- A constructor without fields only can be returned untouched:
+
+instance Functor U1 where
+  fmap _ U1 = U1
+
+-- A field that is not ‘p’ parameter should not change:
+
+instance Functor (K1 i c) where
+  fmap _ (K1 a) = a
+
+-- Metadata has no effect, just unwrap it and continue with the inner value,
+-- if the inner value is an instance of ‘Functor’:
+
+instance Functor f => Functor (M1 i c f) where
+  fmap f (M1 a) = M1 (fmap f a)
+
+-- When we have a sum, we should try to map what we get, provided that it
+-- contains something that has ‘Functor’ instance:
+
+instance (Functor l, Functor r) => Functor (l :+: r) where
+  fmap f (L1 a) = L1 (fmap f a)
+  fmap f (R1 a) = R1 (fmap f a)
+
+-- The same for products:
+
+instance (Functor a, Functor b) => Functor (a :*: b) where
+  fmap f (a :*: b) = fmap f a :*: fmap f b
+```
+
+## Example: counting constructor fields
+
+I think now we're dangerous enough to implement a simple (and pretty
+useless) type class that will count constructor fields of a given value.
+(Yeah, I could go with the venerable `Encode` example here, but I like to be
+“original”, so you get a dumber example instead :-D)
+
+The type class looks like this:
+
+```haskell
+class CountFields a where
+  -- | Return number of constuctor fields for a value.
+  countFields :: a -> Natural
+```
+
+We will start by implementing `countFields` method that works on
+representations:
+
+```haskell
+instance CountFields (V1 p) where
+  countFields _ = 0
+
+instance CountFields (U1 p) where
+  countFields _ = 0
+
+instance CountFields (K1 i c p) where
+  countFields _ = 1
+
+instance CountFields (f p) => CountFields (M1 i c f p) where
+  countFields (M1 x) = countFields x
+
+instance (CountFields (a p), CountFields (b p)) => CountFields ((a :+: b) p) where
+  countFields (L1 x) = countFields x
+  countFields (R1 x) = countFields x
+
+instance (CountFields (a p), CountFields (b p)) => CountFields ((a :*: b) p) where
+  countFields (a :*: b) = countFields a + countFields b
+```
+
+Let write a single function called, say, `defaultCountFields` that does the
+whole thing:
+
+```haskell
+defaultCountFields :: (Generic a, CountFields (Rep a)) => a -> Natural
+defaultCountFields = countFields . from
+```
+
+But here is the catch, the code above does not compile. The `CountFields`
```

Comment:
Constraints have the kind `Constraint` :-) For example `show :: Show a => a -> String`, where the `Show a` thing is not something magical, you can manipulate it too, for example you can get it from a type function. So it needs a kind.

Type class does not have a kind per se, but every type class has a corresponding constraint (like `Show` type class has `Show :: * -> Constraint`, because we need a way to express the idea of constraint and it should be well-kinded, so here for example it says give me a type and I'll give you a constraint that will require that type to have an instance of `Show` type class).